### ISSUE TITLE
umask setting is crazy

### ISSUE THREAD

**Issue by ayghor** [voidlinux/void-packages#2994]
It recently came to my attention that my default umask value was
magically changed to 022, when permissions got messed up and another
user couldn't group-write my files.

I tried to change login.defs, but It didn't work, because it was being
overriden by /etc/profile. Running `git log --all --grep='umask'`, I
noticed it was changed a lot and that it's state is confusing (not to
say inconsistent):

 * /etc/login.defs	-> UMASK 077
 * /etc/profile		-> umask 022
 * /etc/pam.d/su	-> umask=022

IMHO, I think that default umask should be set by, and only by, PAM:

 * usergroup setting
 * shell independence
 * consistency
 * works on non-login shells

and that it should be 077, as suggested by pancake, since home dirs are
world-readable.

Output of `git log --all --grep='umask'`:

    commit 915e2fa0af97aa2403f4cb2753db002b22f39a41
    Author: Juan RP <xtraeme@voidlinux.eu>
    Date:   Thu Jul 2 08:06:52 2015 +0200
    
        util-linux: set su(1) default umask via pam.
        
        This should override umask setting from /etc/login.defs to 022.
        
        Should help #1941
    
    commit 83b38fc69fbf1812038689aedc7c75b848d97ebe
    Author: Juan RP <xtraeme@voidlinux.eu>
    Date:   Sat Jun 6 10:58:35 2015 +0200
    
        shadow: set umask to 077 via login.defs for sane home user dirs.
    
    commit 64c332096b245420531f4321509a2bcb0aee1473
    Author: Juan RP <xtraeme@voidlinux.eu>
    Date:   Sat Jun 6 10:57:30 2015 +0200
    
        base-files: set umask 022 via /etc/profile for sane defaults.
    
    commit a7b6662df99be1f39fa3bc1ef29479bc8183da0a
    Author: Juan RP <xtraeme@voidlinux.eu>
    Date:   Sat Jun 6 08:38:42 2015 +0200
    
        shadow: set default umask in login.defs... fixes weird perms introduced yesterday.
    
    commit 284b8a98d60c76c4d984b06f0f5315272c190370
    Author: Juan RP <xtraeme@voidlinux.eu>
    Date:   Fri Jun 5 11:26:55 2015 +0200
    
        shadow: set umask 077 in /etc/login.defs to avoid world-readable home dirs.
        
        Thanks to pancake for finding this.
    
    commit 9eb2562b3be1140d3c838561903ffb226623952a
    Author: Juan RP <xtraeme@gmail.com>
    Date:   Sun Jan 11 11:44:32 2015 +0100
    
        util-linux: enable pam_umask(8) for su(1).
        
        This way the su(1) session will always have a proper umask(2),
        set from /etc/login.defs by default.
    
    commit 54357100cc489bdd73efe5c23b0d8a1398074fa2
    Author: Juan RP <xtraeme@gmail.com>
    Date:   Sun Jan 11 11:03:49 2015 +0100
    
        xbps: merge patch from master to setup a sane umask(2) while unpacking.
    
    commit 31aa62bab8eebf040aa9bf51dbdf5e125421b102
    Merge: e171e5a f338144
    Author: Juan RP <xtraeme@gmail.com>
    Date:   Thu Dec 25 19:23:47 2014 +0100
    
        Merge pull request #773 from ayghor/master
        
        Use pam_umask(8) instead of /etc/profile.
    
    commit f338144a019de9f0b9fae66dfaf0f5bc5b7d8cef
    Author: dota? =op <rogi@skylittlesystem.org>
    Date:   Wed Dec 24 10:29:01 2014 -0200
    
        Use pam_umask.so instead of /etc/profile.
        
        Instead of setting teh default umask at /etc/profile, let pam_umask.so
        do it's job. Also allow usergroups!

PS: thank you, pancake!

**Comment by xtraeme** [comment]
```
$ sudo useradd -m xtraeme
$ ls -ld /home/xtraeme
drwx------ 2 xtraeme xtraeme 4096 nov 17 09:11 /home/xtraeme
```

Wrong, `/etc/login.defs` setting is only applied to `useradd(8)` and then `/etc/profile` and `pam` set `umask` to `022` for a sane default.

**Comment by ayghor** [comment]
No, I'm not! Teh manpage is on my side, and I just tested it myself.

1 - remove `umask 0xx` in /etc/profile
2 - remove `umask=0xx` option from pam_umask in /etc/pam.d/* files
3 - set UMASK 0xx in /etc/login.defs

and now logging in on a tty, changing user on `su` or `su -l` gives teh
correct result:

 * 0xx for root
 * 00x for normal users

A bit of teh manpage:

    NAME
           pam_umask - PAM module to set the file mode creation mask
    
    SYNOPSIS
           pam_umask.so [debug] [silent] [usergroups] [umask=mask]
    
    DESCRIPTION
           pam_umask is a PAM module to set the file mode creation mask of the
           current environment. The umask affects the default permissions assigned
           to newly created files.
    
           The PAM module tries to get the umask value from the following places
           in the following order:
    
               •umask= argument
    
               •umask= entry in the user's GECOS field
    
               •UMASK= entry from /etc/default/login
    
               •UMASK entry from /etc/login.defs

    ... 

If home dirs are not world-readable, I think umask 022 is ok... sorry,
pancake!

**Comment by ayghor** [comment]
Also, from `man login.defs`:

    ...

           UMASK (number)
               The file mode creation mask is initialized to this value. If not
               specified, the mask will be initialized to 022.
    
               useradd and newusers use this mask to set the mode of the home
               directory they create
    
               It is also used by login to define users' initial umask. Note that
               this mask can be overridden by the user's GECOS line (if
               QUOTAS_ENAB is set) or by the specification of a limit with the K
               identifier in limits(5).
    
    ...

**Comment by ayghor** [comment]
I just realized that by having 022 umask, useradd also makes user homes world-readable. Perhaps we should set this in /etc/pam.d/system-login and /etc/pam.d/common-session

**Comment by chneukirchen** [comment]
IMHO, umask should be set to 022 and useradd should be patched to default to more restrictive umask...

OTOH if it can be set in common-session, perhaps that is good.

**Comment by ayghor** [comment]
I'm looking into this!

**Comment by ayghor** [comment]
Matter of fact, there exists a

           pam_mkhomedir - PAM module to create users home directory

**Comment by ayghor** [comment]
After looking around, I found some solutions to this:

# 1 create user dirs with default umask

this is sane, IMHO. On most distros I remember, teh default is to have a
`umask 022` and world-readable home directories. I think it is what I and
most people would expect.


# 2 patch `useradd`

By looking at `useradd.c`, my conclusion is that it will _only_ obey
`/etc/login.defs` and setting `pam_umask.so` in `/etc/pam.d/useradd` is
completely useless.

Two patches can be proposed:


## 2.1 new option

to `chmod` teh newly created user directory. A new `UMASK` setting could
sit in `/etc/defaults/useradd`


## 2.2 obey `pam_umask.so`

so we can override `umask` in `/etc/pam.d/useradd`


# 3 don't create user dirs with `useradd`

instead, let `pam_mkhomedir.so` create user directories when they first
login. Then we can set teh `umask` option.


# 4 be lazy

 file                               | setting
 -----------------------------------|---------------------------------------------------
 `/etc/login.defs`                  | UMASK 077
 `/etc/pam.d/common-auth`           | session required pam_umask.so usergroups umask=022
 `/etc/pam.d/{system-auth,su,su-l}` | session include common-auth

but there could be some cases teh settings wouldn't cover.


# Conclusion

IMHO, best solution is both 2.2 and 2.3.  But I'm submitting 4 as a
temporary solution since I'm not comfy about patching shadow without
caution (and because I'm lazy).

Please, no `umask 022` in `/etc/profile`!

**Comment by ayghor** [comment]
Please, note that this is not just an Idea! Teh current behaviour does
not respect usergroups policy!

**Comment by GReagle** [comment]
For the benefit of anyone else who wants to change the umask, I am documenting what I did to change the umask.  I chose 027.  I changed it in the following files and it seems to work correctly now, for logins into both a virtual terminal (/dev/tty1 and so on) and an XFCE session.  Note that my login shell is fish which does not call /etc/profile.  Note also that I don't know whether creating new users' home directories is affected.  Hope this helps.

```
greg@d610 ~> grep -i umask /etc/pam.d/* /etc/profile
/etc/pam.d/su:session		optional 	pam_umask.so 	usergroups umask=027
/etc/pam.d/su-l:session		optional 	pam_umask.so 	usergroups umask=027
/etc/pam.d/system-auth:session   optional  pam_umask.so    usergroups umask=027
/etc/profile:# Set default umask
/etc/profile:umask 027
```

It was already 077 as shown here which I did not change
```
greg@d610 ~> grep -i umask /etc/login.defs 
#	UMASK		Default "umask" value.
UMASK			077
```
